### PR TITLE
[SofaBaseTopology] Major Change in Topology Containers

### LINKED ISSUE #746: Sofa crashing when the collision mesh file can not be loaded
Hi,
In one of my scenes, I have once mis spelled 'src' in the line code adding a loader for the mesh of an object's collision model:
`collision.createObject('Mesh', name="mesh", scr="@loader")`.
As a result, when animating the scene, and at the moment when a collision is detected, Sofa shuts down while no error message is provided in the terminal.
I have attached the code file of a scene that reproduces the error, using the Sphere and the Floor prefabs of stlib.physics.rigid

```python
from stlib.scene import ContactHeader, MainHeader
from stlib.physics.rigid import Floor
from stlib.solver import DefaultSolver


def createScene(rootNode):

	MainHeader(rootNode, gravity=[0.0,-9810.0,0.0])	
	ContactHeader(rootNode, alarmDistance=5, contactDistance=2)

	ball = rootNode.createChild("Floor")

	totalMass = 1.0
	volume = 1.0
	inertiaMatrix = [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0]

	ball.createObject('MechanicalObject', name="DOF", template="Rigid", translation=[0.0,0.0,0.0], rotation=[0.0,0.0,0.0])
	ball.createObject('UniformMass', name="mass", mass=[totalMass, volume, inertiaMatrix[:]])

	ball.createObject('UncoupledConstraintCorrection')
	ball.createObject('EulerImplicit', name='odesolver')
	ball.createObject('CGLinearSolver', name='Solver')

	visual = ball.createChild("Floor Visual")
	visual.createObject('OglModel', name="Visual", fileMesh="ball.obj", color=[0.1,0.0,1.0])
	visual.createObject('RigidMapping')

	# collision model for the ball

	collision=ball.createChild("FloorCollisionModel")
	collision.createObject('MeshObjLoader', name='loader', filename="ball.obj", triangulate="true")

	#line code responsible for the crash (scr is mispelled)
	collision.createObject('Mesh', name="mesh", scr="@loader")
	# by writing src correctly, the scene works well 

	collision.createObject('MechanicalObject')

	collision.createObject('Triangle')
	collision.createObject('Line')
	collision.createObject('Point')

	collision.createObject('RigidMapping')
	# end collision


	Floor(rootNode, translation=[0.0,0.0,0.0],
                  uniformScale=10.0,
                  totalMass=0.0015, isAStaticObject=True)


	return rootNode

````

[LoaderSRCcrash.zip](https://github.com/sofa-framework/sofa/files/2257023/LoaderSRCcrash.zip)

______________________________________________________

Suggested labels:

### LINKED ISSUE #955: TetrahedralCorotationalFEMForceField has a bad behavior when removing tetrahedron
Seems to be a problem of force recomputation. When removing some elements, other start to twist..

______________________________________________________

Suggested labels:

### LINKED ISSUE #956: Several Methods in Topology are emitting false negative msg_error
See all those scenes that are working in fact:
https://ci.inria.fr/sofa-ci-dev/job/sofa-framework/job/master/CI_CONFIG=ubuntu_clang-3.8,CI_PLUGINS=options,CI_TYPE=debug/691/warnings3Result/

______________________________________________________

Suggested labels:

### PR DESCRIPTION
**Pull request by epernod**
- **Major change**:
Right now in Edge/Triangle/Quad/Tetrahedron/Hexahedron TopologyContainer, by default all cross element buffers like:
`EdgesInTriangles `or `TrianglesAroundEdges `as well as `d_triangles `array inside a `TetrahedronToplogy `where empty at start and created only the first time a getter is called.
For example: `getEdgesInTriangleArray()` was calling: `if(m_edgesInTriangleArray.empty()) createEdgesInTriangleArray()` 

This works but means at start component are not fully init (potential unstable behavior) and there is no warranty at which timestep buffer will be created. 
Furthermore means that getter are not const and could change the content of the container which is a bad design (in my opinion)

**This PR force the computation of all buffers at component init** using a dedicated method `initTopology()` which is called in cascade -> `Tetrahedron::initTopology()` will call `Triangle::initTopology()` -> `Edge::initTopology()`
**and remove all buffer creation call from the getter.**

- Additional changes: 
   - Add more check in the buffer creation methods. Potentially exit if creation is not possible (will return bool in future changes).
   - Always check buffer size access in the getter. If out of bound acces, returns Invalid element and only emit a warning if CHECK_TOPOLOGY is ON, as this call could be logical and the caller should check the invalid return. 
   - Update all topologicalMapping to init the target topology at end of mapping init. 
   - **Add some static invalid buffers in sofa::core::topology::Topology.h -> Im not fully happy about those add. Any better suggestion is welcomed.**

fixes #746
fixes #955
fixes #956 
______________________________________________________

This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings.
- [x] does not generate new unit test failures.
- [x] does not generate new scene test failures.
- [x] does not break API compatibility.
- [x] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

**Comment by guparan** [comment]
[ci-build][with-all-tests]

File: SofaKernel/modules/SofaBaseTopology/TetrahedronSetTopologyContainer.cpp [hunk 1]
```diff
@@ -231,22 +237,22 @@ void TetrahedronSetTopologyContainer::createEdgesInTetrahedronArray()
                     }
                 }
 
-				if (CHECK_TOPOLOGY)
-					if (foundEdge==false)
-						msg_warning() << "[TetrahedronSetTopologyContainer::getTetrahedronArray] cannot find edge for tetrahedron " << i << "and edge "<< j;
-
+                if (CHECK_TOPOLOGY && foundEdge==false)
+                    msg_warning() << "[TetrahedronSetTopologyContainer::getTetrahedronArray] cannot find edge for tetrahedron " << i << "and edge "<< j;
```
Comment (tgaugry) [inline SofaKernel/modules/SofaBaseTopology/TetrahedronSetTopologyContainer.cpp]: `msg_warning_when`

**Comment by damienmarchal** [comment]
Sorry @epernod  I didn't noticed that you looked at it. I was focused on the open issue.

**Comment by epernod** [comment]
Should fix issues: #746, #956 and probably #955

**Comment by damienmarchal** [comment]
This is ready ???? So I merge it. :)

**Pull request merged**